3 state rotation in analysis_v2

stephlazar opened the pull request:
Adds three_state_rotation.py which rotates averaged data based on cal states for g, e, f.
Adds this functionality to MultiQubit_TimeDomain_Analysis.
Modifies the get_rotations method of CalibrationPoints to not enforce_two_cal_states anymore.

Before (i,e. with two-state rotation):
![Rabi_qb1_20200423_094604](https://user-images.githubusercontent.com/14844492/81197470-997fbe00-8fc0-11ea-84fd-33259a1e022a.png)

With this pull request (i.e. with 3-state rotation):
![Rabi_qb1_20200423_094604](https://user-images.githubusercontent.com/14844492/81197556-b916e680-8fc0-11ea-8fac-5bff82d054a7.png)

FYI @antsr @chellings

Commit:
Adds three_state_rotation.py which rotates averaged data based on cal states for g, e, f.
Adds this functionality to MultiQubit_TimeDomain_Analysis.

Commit:
Changes get_rotations in CalibrationPoints to not enforce_two_cal_states.
Fixes bug in call to predict_proba_avg_ro in rotate_data_3_cal_states.

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

nathlacroix commented on pycqed/measurement/calibration_points.py:
```diff
@@ -107,7 +107,8 @@ def _check_qb_names(self, qb_names):
-    def get_rotations(self, last_ge_pulses=False, qb_names=None):
+    def get_rotations(self, last_ge_pulses=False, qb_names=None,
+                      enforce_two_cal_states=False):
         """
```
could you maybe add to the docstring in which cases this new flag is intended to be used? Not sure I really understand in which context we would need it

nathlacroix replied:
Okay, rereading your pull request I guess this was used before when we had 3 calib points but wanted only 2 for the rotation. Would be great to still specify it in the docstring.

stephlazar replied:
Makes sense, I will add it to docstring.

nathlacroix commented on pycqed/analysis/three_state_rotation.py:
```diff
@@ -0,0 +1,146 @@
+    Predicts probabilities of each state of average readout data
+    X using calibration points for 3 (fixed number) of calibration points
+    Args:
+        X: (n_samples, n_channels) average readout data points.
```
Docstring is outdated: there is no more "proj_pt_on_triangle".
I would to the docstring that probabilities returned can be negative and larger than 1. The mapping to the [0,1] interval would require the code that I had earlier, or a smarter way that does not depend on the shapely package. Do we think this is necessary/woth to concentrate on for the medium term? if yes, I could create a git issue for it (and let this code go as it is for now). @stephlazar  what do you think?

stephlazar replied:
I think this would definitely be nice to have since after this pull request, all our measurements with f-level will use this analysis.
Maybe it makes sense that you implement it in this pull request and we merge it only when that is done. What do you think?

nathlacroix replied:
I would rather do the opposite, let's merge this one for now to move on and I will create a git issue to remember we have to tackle this issue but it's not top 1 on the priority list right now, I believe. But it would be great to be able to use the code in this pull request already in the meanwhile.

nathlacroix replied:
see issue #75 .

nathlacroix reviewed (approved):
Thanks for incorporating the code in pycqed :)

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Improves MultiQubit_TimeDomain_Analysis to do cal states rotation even when different
  number of cal states are specified for the qubits. Example, it is not able to rotate
  qb1 based on g, e, f cal states, and qb2 based on only g, e cal states.
Updates the metadata information in measure_cphase to make sure of this new feature.

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

Commit:
Merge branch 'qudev_master' into enh/3_state_rotation

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -431,158 +432,243 @@ def get_cal_data_points(self):
-    def rotate_data(meas_results_per_qb, channel_map,
+    def rotate_data(qb_name, meas_results_per_qb, channel_map,
                     cal_states_dict, data_to_fit):
+        print('in 2state rotation')
```
debug print statement

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -431,158 +432,243 @@ def get_cal_data_points(self):
+    @staticmethod
+    def rotate_data_3_cal_states_TwoD(qb_name, meas_results_per_qb,
+                                      channel_map, cal_states_dict):
+        print('in 3state rotation 2D')
```
debug print statement.

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -431,158 +432,243 @@ def get_cal_data_points(self):
+    @staticmethod
+    def rotate_data_TwoD(qb_name, meas_results_per_qb, channel_map,
+                         cal_states_dict, data_to_fit):
+        print('in 2state rotation 2D')
```
debug

nathlacroix commented:
@stephlazar  good to merge for me :)

nathlacroix committed:
Clean up print and log messages in timedomain_analysis
